### PR TITLE
React 17 compatibility

### PR DESCRIPTION
**Pull request by dusave**
This PR:
- Updates the `react` lifecycles to be `v17` compatible
- Updates `babel-*@6` to `@babel/*@7`
- Updates other outdated dependencies

### CODE REVIEW

File: validate.js
Comment (decompil3d) [inline validate.js]: Why does this need to be copied to `state`?

Comment (dusave) [reply validate.js]: In the original code, we were comparing `this.props.validate === nextProps.validate` to determine if we should revalidate. Since we no longer have access to the previous props in `getDerivedStateFromProps`, I thought this would be the way around that. I also considered `componentDidUpdate` as a potential alternative, however that doesn't get executed on the initial mount.

**Review by decompil3d — APPROVED** [review]

File: package-lock.json
Comment (3rd-Eden) [inline package-lock.json]: Use NPM registry

**Review by 3rd-Eden — CHANGES_REQUESTED** [review]

File: demo/package.json
Comment (indexzero) [inline demo/package.json:11]: Recently someone turned me on to the fact you should just be able to do:

```
"lint": "eslint ."
```

File: package.json
Comment (indexzero) [inline package.json:31]: Recently someone turned me on to the fact you should just be able to do:

```
"lint": "eslint ."
```

File: validate.js
Comment (indexzero) [inline validate.js]: I am pretty sure this will not work because `getDerivedStateFromProps` is called multiple times in the component's lifecycle, not simply when it receives props. 

[This post](https://blog.logrocket.com/the-new-react-lifecycle-methods-in-plain-approachable-language-61a2105859f3) outlines the new lifecycle well. [This other post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) goes into details about best practices for `getDerivedStateFromProps`. 

Perhaps we can accomplish this without `getDerivedStateFromProps`?

File: validate.js
Comment (indexzero) [inline validate.js:80]: Why change this?

File: validates.js
Comment (indexzero) [inline validates.js:62]: Why change this?

**Review by indexzero — CHANGES_REQUESTED** [review]
Would like to understand the changes a bit more – I suspect we probably need additional unit test coverage since `getDerivedStateFromProps` has a different call pattern than `componentWillReceiveProps` did (i.e. it is called more frequently).

File: validate.js
Comment (dusave) [reply validate.js:80]: `componentDidMount` is called after render, whereas `componentWillMount` is called before render, so I updated the JSDoc to reflect that. I made this change since `componentWillMount` is being deprecated.

File: package.json
Comment (dusave) [reply package.json:31]: This would cause it to also lint against the `lib/` directory. Should we go with the exclusion route rather than the inclusion route?

File: validate.js
Comment (dusave) [reply validate.js]: I may be completely incorrect in my thinking here, but from what I understand from [this lifecycle diagram](https://twitter.com/dan_abramov/status/981712092611989509?lang=en) from Dan Abramov, it appears that `getDerivedStateFromProps` is called on initial mount and when props have changed. Based on that knowledge, the reason I went with this approach was so that we could check if the prop value had changed and if not, not update state nor re-render. 

Alternative approaches considered were:
- Check in `shouldComponentUpdate`, which I believe would be called more often than `getDerivedStateFromProps` since it happens on both `state` _and_ `prop` changes.
- Use `componentDidUpdate` in conjunction with the `constructor` to check, but (if I understand correctly) this happens post-render which, in the event of a change, would cause a re-render.

All that said, perhaps none of this is needed. I simply attempted a migration, but do we anticipate the `validate` function prop to change after the initial mount? If not, we can dispense with this entire block of code. Thoughts?

File: package-lock.json
Comment (dusave) [reply package-lock.json]: Forgot to swap out my `.npmrc`. Fixed 👍

File: validate.js
Comment (indexzero) [reply validate.js]: We must go deeper. 

The post you linked to is the lifecycle diagram circa `react@16.3` – but `react@16.4` introduced the lifecycle diagram you can find [here](https://medium.com/@_ChrisBradshaw/blogdidmount-2018-guide-to-react-component-lifecycle-methods-1614e0bbe80a).

react-164-lifecycle

As you can see the difference is that you'll get calls to `getDerivedStateFromProps` on calls to `setState` – this extra call is where I believe this logic will break down.

Comment (dusave) [reply validate.js]: Ah, I see the issue now. I'll think on that a bit more and come up with a different approach.

Comment (dusave) [reply validate.js]: Also, I need to correct my previous comment:

- `componentDidUpdate` could potentially work here if we utilize the same `setState` call:
> React ensures that any setState calls that happen during componentDidMount and componentDidUpdate are flushed before the user sees the updated UI.

Given that additional information, I'll try that path.

Comment (decompil3d) [reply validate.js]: @dustinsavery -- did you read [the second article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) that @indexzero mentioned? At the bottom, it covers alternatives to `getDerivedStateFromProps` for memoization cases like the one here.

File: validate.js
Comment (indexzero) [reply validate.js:80]: That makes sense. I forgot that `componentWillMount` was one of the lifecycle methods being deprecated.

File: validate.js
Comment (dusave) [reply validate.js]: I've run a number of tests for each of the options and here is what I've found:
- `componentDidUpdate`: By checking `prevProps.validate !== this.props.validate` and executing the function to update the validity, I was able to accomplish the expected behavior. When the `validate` function is swapped, it calls `componentDidUpdate`, re-evaluates the validity, and updates state. And since this is behind the above if check, it doesn't execute unless the function has changed. However, to accomodate this behavior (in conjuction with [this current logic](https://github.com/godaddy/react-validation-context/blob/master/validate.js#L111-L115)), I extracted out the current logic into a new function to keep it DRY. This seems to better than the current approach as it accomplishes the same thing, but without tracking the function in state.
- `getDerivedStateFromProps` (current implementation in PR): This was called on every update cycle (as expected per the diagram). However, since we do an early abort if the method hasn't changed, it didn't cause an increase in render cycles. The downside with this approach is we have to track the function in state, which is not ideal.
- `memoize`/`componentDidUpdate`: I saw no noticeable difference in performance or number of calls with this method from the plain `componentDidUpdate` option.

Given these findings, unless anyone objects, I think going with option 1 makes the most sense.

File: test/validate.test.js
Comment (decompil3d) [inline test/validate.test.js]: Since this runs in node, you can use `setImmediate` instead. ([relevant thread](https://stackoverflow.com/questions/24117267/nodejs-settimeoutfn-0-vs-setimmediatefn))

File: validate.js
Comment (decompil3d) [inline validate.js]: You can pass the function straight in:

```js
this.setState({ validates: this.props.validate(this.state.valids) }, executeOnValidChange);
```

**Review by decompil3d — CHANGES_REQUESTED** [review]

File: test/validate.test.js
Comment (dusave) [reply test/validate.test.js]: That's the function I was trying to remember, thank you! I knew there was a way, but couldn't recall it.

**Review by decompil3d — APPROVED** [review]
Please just make sure that this is properly messaged in the CHANGELOG, since it is ***not*** full React 17 compat yet, as this still uses the legacy context API. You'll need to update this project to use the new context API too (which is another ticket anyway) and that will likely make this a major bump once done. But this PR is fine for now.

It is a major bump since it requires a higher version of react as a peer dependency as well as `babel@7`.

**Comment by dusave** [comment]
@decompil3d Should we merge this (once approved) into a `react-17` branch and wait until the context upgrades are complete before cutting a new version?

**Comment by decompil3d** [comment]
@dustinsavery I think that would be alright. @indexzero thoughts?

**Review by 3rd-Eden — APPROVED** [review]

**Review by indexzero — APPROVED** [review]
Did some serious spelunking into the tests and it appears that using `componentDidUpdate` ensures the same cardinality of `render` invocations – [logs are here](https://gist.github.com/indexzero/81c33837f186e6dd8f5502bb88195d9c) for those who are interested.

**Pull request merged**
